File: src/python/meshioplusplus/_voxelize.py
"""Voxelization (``meshioplusplus.voxelize``).

Build a regular grid around a mesh and, optionally, keep only the cells its
surface passes through or encloses.

The output is an **ordinary** :class:`~meshioplusplus._mesh.Mesh` -- one
``hexahedron`` cell block over a shared corner lattice -- and that is the whole
design. Once it is a mesh, every writer already handles it, ``view``/``screenshot``
already render it, ``crop``/``split``/``data calc`` already operate on it,
``--color-by`` already colours it and ``isosurface`` already contours whatever
field it carries. None of that needed new code.

Three fill rules:

``all``
    Every cell of the (optionally padded) bounding box. The mesh contributes only
    its extent, so this is a background grid rather than a shape.
``surface``
    Only cells a surface triangle passes through, by exact triangle/box overlap.
    Needs no closed surface, so it works on an open sheet.
``inside``
    Only cells whose centre is inside the surface. Needs a sign, and therefore a
    surface closed enough for the chosen convention.

This reference implementation is a byte-for-byte twin of the C++ core, pinned by
``tests/python/test_voxelize.py::test_cpp_matches_python``.

Public API:

* :func:`voxelize` -- build a grid around a mesh.
"""

from __future__ import annotations

import numpy as np

from ._grid import DEFAULT_MAX_CELLS, _lattice_py
from ._mesh import Mesh
from ._sdf import _sample_py, _soup

__all__ = ["voxelize"]

_PREFIX = "meshio++: voxelize: "
#: The same string, under a name the shared resolver can shadow safely.
_VOX_PREFIX = _PREFIX
_FILLS = ("all", "surface", "inside")


def _tri_box_overlap(centre, half, a, b, c):
    """Exact triangle / axis-aligned-box overlap by the separating axis theorem.

    Transcribed from ``detail/tri_box.hpp``: thirteen axes, in the same order.
    A bounding-box test is not this -- a long diagonal triangle overlaps far more
    boxes than it enters.
    """
    v = np.array([a - centre, b - centre, c - centre])
    e = np.array([v[1] - v[0], v[2] - v[1], v[0] - v[2]])

    def separated(p, rad):
        return p.min() > rad or p.max() < -rad

    for k in range(3):
        ex, ey, ez = abs(e[k][0]), abs(e[k][1]), abs(e[k][2])
        p = -e[k][2] * v[:, 1] + e[k][1] * v[:, 2]
        if separated(p, ez * half[1] + ey * half[2]):
            return False
        p = e[k][2] * v[:, 0] - e[k][0] * v[:, 2]
        if separated(p, ez * half[0] + ex * half[2]):
            return False
        p = -e[k][1] * v[:, 0] + e[k][0] * v[:, 1]
        if separated(p, ey * half[0] + ex * half[1]):
            return False

    for k in range(3):
        if separated(v[:, k], half[k]):
            return False

    normal = np.cross(e[0], e[1])
    vmin = np.where(normal > 0.0, -half - v[0], half - v[0])
    vmax = np.where(normal > 0.0, half - v[0], -half - v[0])
    if np.sum(normal * vmin) > 0.0:
        return False
    return np.sum(normal * vmax) >= 0.0


def _resolve_bounds(mesh, bounds, padding, padding_relative, _PREFIX=None):
    _PREFIX = _VOX_PREFIX if _PREFIX is None else _PREFIX
    if bounds is not None:
        b = np.asarray(bounds, dtype=np.float64).reshape(-1)
        if b.size != 6:
            raise ValueError(f"{_PREFIX}bounds must be (xlo, ylo, zlo, xhi, yhi, zhi)")
        lo, hi = b[:3].copy(), b[3:].copy()
        for k in range(3):
            if not hi[k] >= lo[k]:
                raise ValueError(
                    f"{_PREFIX}bounds are inverted on axis {k} "
                    f"(lo {lo[k]} > hi {hi[k]})"
                )
    else:
        pts = np.asarray(mesh.points, dtype=np.float64)
        if len(pts) == 0:
            raise ValueError(
                f"{_PREFIX}the mesh has no points, so it has no bounding box to cover "
                "(pass explicit bounds)"
            )
        if pts.shape[1] == 2:
            pts = np.column_stack([pts, np.zeros(len(pts))])
        lo, hi = pts.min(axis=0), pts.max(axis=0)

    diag = float(np.sqrt(np.sum((hi - lo) ** 2)))
    pad = padding + padding_relative * diag
    if pad < 0.0:
        raise ValueError(f"{_PREFIX}padding is negative")
    return lo - pad, hi + pad


def _resolve_lattice(
    mesh,
    resolution,
    cell_size,
    bounds,
    padding,
    padding_relative,
    max_cells,
    _PREFIX=None,
):
    """The numpy twin of ``detail::lattice_resolve``.

    The prefix is a parameter for the same reason it is in C++: ``compute_sdf``
    resolves the identical six fields the identical way, and its errors must name
    ``sdf`` rather than ``voxelize``.
    """
    _PREFIX = _VOX_PREFIX if _PREFIX is None else _PREFIX
    if (resolution is None) == (cell_size is None):
        raise ValueError(f"{_PREFIX}give exactly one of resolution and cell_size")
    lo, hi = _resolve_bounds(mesh, bounds, padding, padding_relative, _PREFIX)

    if resolution is not None:
        dims = np.asarray(resolution, dtype=np.int64).reshape(-1)
        if dims.size != 3:
            raise ValueError(f"{_PREFIX}resolution must be three cell counts")
        for k in range(3):
            if dims[k] <= 0:
                raise ValueError(
                    f"{_PREFIX}resolution must be positive on every axis, got "
                    f"{int(dims[k])} on axis {k}"
                )
        spacing = np.array(
            [(hi[k] - lo[k]) / float(dims[k]) for k in range(3)], dtype=np.float64
        )
    else:
        cell = float(cell_size)
        if not cell > 0.0:
            raise ValueError(f"{_PREFIX}cell_size must be positive")
        dims = np.zeros(3, dtype=np.int64)
        for k in range(3):
            extent = hi[k] - lo[k]
            dims[k] = int(np.ceil(extent / cell)) if extent > 0.0 else 0
            if dims[k] <= 0:
                raise ValueError(
                    f"{_PREFIX}the bounding box is degenerate on axis {k}, so a cell size "
                    "cannot fill it (pass an explicit resolution or bounds)"
                )
        spacing = np.full(3, cell, dtype=np.float64)

    cells = int(dims[0]) * int(dims[1]) * int(dims[2])
    if max_cells > 0 and cells > max_cells:
        raise ValueError(
            f"{_PREFIX}the requested grid has {cells} cells, above the limit of {max_cells} "
            "(raise max_cells, coarsen the resolution, or use a band)"
        )
    return lo, spacing, dims


def _voxelize_py(
    mesh,
    resolution,
    cell_size,
    bounds,
    padding,
    padding_relative,
    fill,
    attach_occupancy,
    max_cells,
    sign,
):
    origin, spacing, dims = _resolve_lattice(
        mesh, resolution, cell_size, bounds, padding, padding_relative, max_cells
    )
    nx, ny, nz = (int(d) for d in dims)
    ncells = nx * ny * nz

    occupied = None
    if fill == "surface":
        occupied = np.zeros(ncells, dtype=bool)
        _points, _verts, corners, _source = _soup(mesh)
        half = spacing * 0.5
        for t in range(len(corners)):
            a, b, c = corners[t]
            tlo = np.minimum(np.minimum(a, b), c)
            thi = np.maximum(np.maximum(a, b), c)
            lo_i = np.maximum(np.floor((tlo - origin) / spacing), 0).astype(np.int64)
            hi_i = np.minimum(np.floor((thi - origin) / spacing), dims - 1).astype(
                np.int64
            )
            if np.any(hi_i < lo_i):
                continue
            for k in range(lo_i[2], hi_i[2] + 1):
                for j in range(lo_i[1], hi_i[1] + 1):
                    for i in range(lo_i[0], hi_i[0] + 1):
                        cid = (k * ny + j) * nx + i
                        if occupied[cid]:
                            continue
                        centre = origin + (np.array([i, j, k]) + 0.5) * spacing
                        if _tri_box_overlap(centre, half, a, b, c):
                            occupied[cid] = True
    elif fill == "inside":
        if sign == "unsigned":
            raise ValueError(
                f"{_PREFIX}fill 'inside' needs a sign, but sign='unsigned' was given"
            )
        ii, jj, kk = np.meshgrid(
            np.arange(nx), np.arange(ny), np.arange(nz), indexing="ij"
        )
        idx = np.stack(
            [
                ii.transpose(2, 1, 0).reshape(-1),
                jj.transpose(2, 1, 0).reshape(-1),
                kk.transpose(2, 1, 0).reshape(-1),
            ],
            axis=1,
        ).astype(np.float64)
        centres = origin + (idx + 0.5) * spacing
        dist, _cells, _band = _sample_py(mesh, centres, sign, "angle", 0.0)
        occupied = dist < 0.0

    points, conn = _lattice_py(dims, origin, spacing)
    if occupied is None:
        out = Mesh(points, [] if conn is None else [("hexahedron", conn)])
        kept = ncells
    else:
        keep = np.flatnonzero(occupied)
        kept = int(keep.size)
        sub = (
            conn[keep]
            if conn is not None and kept
            else np.zeros((0, 8), dtype=np.int64)
        )
        # Compact the referenced points in ascending order -- surface.cpp's
        # used/remap pattern, so the output does not depend on traversal.
        used = np.unique(sub) if kept else np.zeros(0, dtype=np.int64)
        remap = np.full(len(points), -1, dtype=np.int64)
        remap[used] = np.arange(len(used), dtype=np.int64)
        out = Mesh(
            points[used] if kept else np.zeros((0, 3)),
            [("hexahedron", remap[sub])] if kept else [],
        )
    if attach_occupancy and kept > 0:
        out.cell_data["voxel:occupancy"] = [np.ones(kept, dtype=np.int64)]

    return out, {
        "dims": [int(d) for d in dims],
        "origin": [float(v) for v in origin],
        "spacing": [float(v) for v in spacing],
        "num_occupied": kept,
    }


def voxelize(
    mesh,
    resolution=None,
    cell_size=None,
    bounds=None,
    padding=0.0,
    padding_relative=0.0,
    fill="all",
    attach_occupancy=False,
    max_cells=DEFAULT_MAX_CELLS,
    sign="pseudonormal",
    watertight_check="warn",
    return_report=False,
):
    """Build a regular grid around ``mesh``.

    Exactly one of ``resolution`` and ``cell_size`` must be given: defaulting one
    would silently pick a grid the caller did not choose, and the cost is cubic in
    that choice.
    """
    if fill not in _FILLS:
        raise ValueError(
            f"{_PREFIX}unknown fill '{fill}' (expected one of: {', '.join(_FILLS)})"
        )

    out = None
    report = None
    try:
        from . import _core

        res = _core.voxelize(
            mesh,
            None if resolution is None else [int(v) for v in resolution],
            None if cell_size is None else float(cell_size),
            None if bounds is None else [float(v) for v in bounds],
            float(padding),
            float(padding_relative),
            fill,
            bool(attach_occupancy),
            int(max_cells),
            sign,
            watertight_check,
        )
        out = res["mesh"]
        report = {
            "dims": res["dims"],
            "origin": res["origin"],
            "spacing": res["spacing"],
            "num_occupied": res["num_occupied"],
        }
    except (ValueError, TypeError):
        raise
    except Exception:
        out = None

    if out is None:
        out, report = _voxelize_py(
            mesh,
            resolution,
            cell_size,
            bounds,
            padding,
            padding_relative,
            fill,
            attach_occupancy,
            max_cells,
            sign,
        )

    return (out, report) if return_report else out

File: src/cpp/src/operations/voxelize.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Regular grids: `grid` from nothing, `voxelize` around a mesh. See
// operations/voxelize.hpp for the contract and detail/grid_lattice.hpp for the
// numbering, which this file does not re-derive.
//
// Anonymous-namespace helpers are prefixed `vox_` (the amalgamation concatenates
// every translation unit into one). Note `grid_` was NOT available: it belongs to
// detail/spatial_hash.hpp's grid_quantize, which is a bucket hash rather than a
// mesh, and reusing it would make two unrelated things look related.

// System includes
#include <algorithm>
#include <array>
#include <cmath>
#include <cstring>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

// Project includes
#include "meshioplusplus/operations/voxelize.hpp"
#include "meshioplusplus/detail/grid_lattice.hpp"
#include "meshioplusplus/detail/surface_distance.hpp"
#include "meshioplusplus/detail/tri_box.hpp"
#include "meshioplusplus/detail/region_remap.hpp"
#include "meshioplusplus/log.hpp"
#include "meshioplusplus/parallel.hpp"

namespace meshioplusplus {

namespace {

constexpr const char* kVoxPrefix = "meshio++: voxelize: ";

// The cell budget check for grid(), which has no LatticeRequest to resolve. It
// is a *named* refusal rather than an allocation failure because the failure
// mode it guards is mundane and predictable: 512^3 is 134 million cells and
// ~11.8 GB of points and connectivity, which a user asks for by typo far more
// often than on purpose. `lattice_resolve` applies the identical rule (and the
// identical message) to the request-shaped callers.
void vox_check_budget(std::int64_t Cells, std::int64_t MaxCells, const char* pPrefix) {
    if (MaxCells <= 0 || Cells <= MaxCells)
        return;
    throw std::invalid_argument(std::string(pPrefix) + "the requested grid has " +
                                std::to_string(Cells) + " cells, above the limit of " +
                                std::to_string(MaxCells) +
                                " (raise max_cells, coarsen the resolution, or use a band)");
}

// The lattice a voxelization runs on. Exactly one of resolution and cell size
// must be given: defaulting one of them would silently pick a grid the caller
// did not choose, and for an object whose cost is cubic in that choice that is
// not a kindness. The rule itself lives in detail/grid_lattice.hpp, since
// `compute_sdf` resolves the identical six fields the identical way.
detail::LatticeSpec vox_resolve_lattice(const Mesh& rMesh, const VoxelOptions& rOptions) {
    detail::LatticeRequest req;
    req.mResolution = rOptions.mResolution;
    req.mCellSize = rOptions.mCellSize;
    req.mBounds = rOptions.mBounds;
    req.mPadding = rOptions.mPadding;
    req.mPaddingRelative = rOptions.mPaddingRelative;
    req.mMaxCells = rOptions.mMaxCells;
    return detail::lattice_resolve(rMesh, req, kVoxPrefix);
}

// Which cells the fill rule keeps, as a per-cell flag. An empty result means
// "all of them", which lets the caller skip the subsetting pass entirely.
std::vector<char> vox_occupancy(const Mesh& rMesh, const detail::LatticeSpec& rSpec,
                                const VoxelOptions& rOptions) {
    if (rOptions.mFill == VoxelFill::All)
        return {};

    const std::int64_t ncells = detail::lattice_num_cells(rSpec);
    std::vector<char> occupied(static_cast<std::size_t>(ncells), 0);
    const std::int64_t nx = rSpec.mDims[0];
    const std::int64_t ny = rSpec.mDims[1];

    if (rOptions.mFill == VoxelFill::Surface) {
        // Exact triangle/box overlap, not a bounding-box test: a long diagonal
        // triangle overlaps far more boxes than it enters.
        const detail::TriangleSoup soup =
            detail::build_triangle_soup(rMesh, rOptions.mDistance.mSurfaceRegion);
        const detail::Vec3 half{
            {rSpec.mSpacing[0] * 0.5, rSpec.mSpacing[1] * 0.5, rSpec.mSpacing[2] * 0.5}};
        // Serial over triangles, each touching only the voxels of its own
        // bounding box: a parallel pass would race on the shared flag array for
        // no real gain, since the inner box is small.
        for (std::size_t t = 0; t < soup.NumTriangles(); ++t) {
            const detail::Vec3& a = soup.mCorners[t * 3 + 0];
            const detail::Vec3& b = soup.mCorners[t * 3 + 1];
            const detail::Vec3& c = soup.mCorners[t * 3 + 2];
            std::array<std::int64_t, 3> lo{}, hi{};
            bool skip = false;
            for (std::size_t k = 0; k < 3; ++k) {
                if (!(rSpec.mSpacing[k] > 0.0)) {
                    skip = true;
                    break;
                }
                const double tlo = std::min(a[k], std::min(b[k], c[k]));
                const double thi = std::max(a[k], std::max(b[k], c[k]));
                double flo = std::floor((tlo - rSpec.mOrigin[k]) / rSpec.mSpacing[k]);
                double fhi = std::floor((thi - rSpec.mOrigin[k]) / rSpec.mSpacing[k]);
                flo = flo < 0.0 ? 0.0 : flo;
                const double last = static_cast<double>(rSpec.mDims[k] - 1);
                fhi = fhi > last ? last : fhi;
                if (fhi < flo) {
                    skip = true;
                    break;
                }
                lo[k] = static_cast<std::int64_t>(flo);
                hi[k] = static_cast<std::int64_t>(fhi);
            }
            if (skip)
                continue;
            for (std::int64_t k = lo[2]; k <= hi[2]; ++k)
                for (std::int64_t j = lo[1]; j <= hi[1]; ++j)
                    for (std::int64_t i = lo[0]; i <= hi[0]; ++i) {
                        const std::int64_t cid = (k * ny + j) * nx + i;
                        if (occupied[static_cast<std::size_t>(cid)])
                            continue;
                        const detail::Vec3 centre{
                            {rSpec.mOrigin[0] + (static_cast<double>(i) + 0.5) * rSpec.mSpacing[0],
                             rSpec.mOrigin[1] + (static_cast<double>(j) + 0.5) * rSpec.mSpacing[1],
                             rSpec.mOrigin[2] +
                                 (static_cast<double>(k) + 0.5) * rSpec.mSpacing[2]}};
                        if (detail::tri_box_overlap(centre, half, a, b, c))
                            occupied[static_cast<std::size_t>(cid)] = 1;
                    }
        }
        return occupied;
    }

    // Inside: sign the distance at each cell centre. This is the only fill that
    // needs a closed surface, and it says so through the watertight check rather
    // than by quietly producing a hollow result.
    const detail::TriangleSoup soup =
        detail::build_triangle_soup(rMesh, rOptions.mDistance.mSurfaceRegion);
    std::vector<detail::Vec3> centres(static_cast<std::size_t>(ncells));
    parallel_for_bw(static_cast<std::size_t>(ncells), [&](std::size_t c) {
        const std::int64_t g = static_cast<std::int64_t>(c);
        const std::int64_t i = g % nx;
        const std::int64_t j = (g / nx) % ny;
        const std::int64_t k = g / (nx * ny);
        centres[c] =
            detail::Vec3{{rSpec.mOrigin[0] + (static_cast<double>(i) + 0.5) * rSpec.mSpacing[0],
                          rSpec.mOrigin[1] + (static_cast<double>(j) + 0.5) * rSpec.mSpacing[1],
                          rSpec.mOrigin[2] + (static_cast<double>(k) + 0.5) * rSpec.mSpacing[2]}};
    });

    SurfaceDistanceOptions dopts = rOptions.mDistance;
    if (dopts.mSign == SdfSign::Unsigned)
        throw std::invalid_argument(std::string(kVoxPrefix) +
                                    "fill 'inside' needs a sign, but sign='unsigned' was given");
    dopts.mBand = 0.0;  // a band would clamp the very sign this fill depends on
    const detail::DistanceQuery query = detail::build_distance_query(soup, dopts);
    const std::vector<detail::DistanceHit> hits = detail::query_distances(query, centres, dopts);
    parallel_for_bw(static_cast<std::size_t>(ncells),
                    [&](std::size_t c) { occupied[c] = hits[c].mSignedDistance < 0.0 ? 1 : 0; });
    return occupied;
}

// Emit only the flagged cells, compacting the points they reference in ascending
// order so the output does not depend on traversal.
Mesh vox_build_subset(const detail::LatticeSpec& rSpec, const std::vector<char>& rOccupied,
                      std::int64_t Kept) {
    Mesh out;
    const std::int64_t nx = rSpec.mDims[0];
    const std::int64_t ny = rSpec.mDims[1];
    const std::int64_t px = nx + 1;
    const std::int64_t py = ny + 1;
    const std::int64_t npoints = detail::lattice_num_points(rSpec);
    const std::int64_t ncells = detail::lattice_num_cells(rSpec);

    // Two passes: mark which points the kept cells reference, then number them in
    // ASCENDING original order. That is surface.cpp's used/remap pattern, and the
    // ascending part is load-bearing rather than incidental -- numbering on first
    // encounter instead would make the point ids depend on the hexahedron's node
    // order, which is a traversal detail no caller should be able to observe (and
    // which the numpy twin would then have to replicate rather than simply sort).
    std::vector<char> used(static_cast<std::size_t>(npoints), 0);
    std::vector<std::int64_t> kept_cells;
    kept_cells.reserve(static_cast<std::size_t>(Kept));
    for (std::int64_t c = 0; c < ncells; ++c) {
        if (!rOccupied[static_cast<std::size_t>(c)])
            continue;
        kept_cells.push_back(c);
        const std::int64_t i = c % nx;
        const std::int64_t j = (c / nx) % ny;
        const std::int64_t k = c / (nx * ny);
        const std::int64_t base = (k * py + j) * px + i;
        const std::int64_t top = base + px * py;
        const std::int64_t nodes[8] = {base, base + 1, base + px + 1, base + px,
                                       top,  top + 1,  top + px + 1,  top + px};
        for (std::int64_t nd : nodes)
            used[static_cast<std::size_t>(nd)] = 1;
    }

    std::vector<std::int64_t> remap(static_cast<std::size_t>(npoints), -1);
    std::int64_t next = 0;
    for (std::int64_t p = 0; p < npoints; ++p)
        if (used[static_cast<std::size_t>(p)])
            remap[static_cast<std::size_t>(p)] = next++;

    std::vector<std::int64_t> conn;
    conn.reserve(static_cast<std::size_t>(Kept) * 8);
    for (std::int64_t c : kept_cells) {
        const std::int64_t i = c % nx;
        const std::int64_t j = (c / nx) % ny;
        const std::int64_t k = c / (nx * ny);
        const std::int64_t base = (k * py + j) * px + i;
        const std::int64_t top = base + px * py;
        const std::int64_t nodes[8] = {base, base + 1, base + px + 1, base + px,
                                       top,  top + 1,  top + px + 1,  top + px};
        for (std::int64_t nd : nodes)
            conn.push_back(remap[static_cast<std::size_t>(nd)]);
    }

    NDArray points =
        NDArray::Uninit(DType::Float64, {static_cast<std::size_t>(next), std::size_t{3}});
    double* pdst = points.As<double>();
    parallel_for_bw(static_cast<std::size_t>(npoints), [&](std::size_t p) {
        const std::int64_t slot = remap[p];
        if (slot < 0)
            return;
        const std::int64_t g = static_cast<std::int64_t>(p);
        const std::int64_t i = g % px;
        const std::int64_t j = (g / px) % py;
        const std::int64_t k = g / (px * py);
        pdst[slot * 3 + 0] = rSpec.mOrigin[0] + static_cast<double>(i) * rSpec.mSpacing[0];
        pdst[slot * 3 + 1] = rSpec.mOrigin[1] + static_cast<double>(j) * rSpec.mSpacing[1];
        pdst[slot * 3 + 2] = rSpec.mOrigin[2] + static_cast<double>(k) * rSpec.mSpacing[2];
    });
    out.AssignPoints(std::move(points));

    if (Kept > 0) {
        NDArray block =
            NDArray::Uninit(DType::Int64, {static_cast<std::size_t>(Kept), std::size_t{8}});
        std::memcpy(block.Data(), conn.data(), conn.size() * sizeof(std::int64_t));
        out.AddCellBlock("hexahedron", std::move(block));
    }
    return out;
}

}  // namespace

VoxelFill voxel_fill_from_name(const std::string& rName) {
    if (rName == "all")
        return VoxelFill::All;
    if (rName == "surface")
        return VoxelFill::Surface;
    if (rName == "inside")
        return VoxelFill::Inside;
    throw std::invalid_argument(std::string(kVoxPrefix) + "unknown fill '" + rName +
                                "' (expected one of: all, surface, inside)");
}

Mesh grid(const std::array<std::int64_t, 3>& rDims, const std::array<double, 3>& rOrigin,
          const std::array<double, 3>& rSpacing, std::int64_t MaxCells) {
    constexpr const char* prefix = "meshio++: grid: ";
    for (std::size_t k = 0; k < 3; ++k) {
        if (rDims[k] < 0)
            throw std::invalid_argument(std::string(prefix) +
                                        "cell counts must not be negative, got " +
                                        std::to_string(rDims[k]) + " on axis " + std::to_string(k));
        if (rDims[k] > 0 && !(rSpacing[k] > 0.0))
            throw std::invalid_argument(
                std::string(prefix) + "spacing must be positive on every axis with cells, got " +
                std::to_string(rSpacing[k]) + " on axis " + std::to_string(k));
    }
    detail::LatticeSpec spec;
    spec.mOrigin = rOrigin;
    spec.mSpacing = rSpacing;
    spec.mDims = rDims;
    vox_check_budget(detail::lattice_num_cells(spec), MaxCells, prefix);
    return detail::lattice_build_mesh(spec);
}

VoxelResult voxelize(const Mesh& rMesh, const VoxelOptions& rOptions) {
    const detail::LatticeSpec spec = vox_resolve_lattice(rMesh, rOptions);

    VoxelResult result;
    result.mDims = spec.mDims;
    result.mOrigin = spec.mOrigin;
    result.mSpacing = spec.mSpacing;

    // Every point and cell of the output is new, so nothing can be remapped --
    // the same situation slice/isosurface/extract_surface are in.
    detail::warn_regions_dropped(rMesh, "voxelize");

    const std::int64_t ncells = detail::lattice_num_cells(spec);
    const std::vector<char> occupied = vox_occupancy(rMesh, spec, rOptions);

    if (rOptions.mFill == VoxelFill::All || occupied.empty()) {
        result.mMesh = detail::lattice_build_mesh(spec);
        result.mNumOccupied = ncells;
        if (rOptions.mAttachOccupancy && ncells > 0) {
            // Every cell is kept, so occupancy is all ones -- constant, but
            // emitted anyway so that a caller can switch fill modes without
            // their downstream pipeline discovering the array has vanished.
            NDArray occ = NDArray::Uninit(DType::Int64, {static_cast<std::size_t>(ncells)});
            std::int64_t* dst = occ.As<std::int64_t>();
            parallel_for_bw(static_cast<std::size_t>(ncells), [&](std::size_t c) { dst[c] = 1; });
            std::vector<NDArray> blocks;
            blocks.push_back(std::move(occ));
            result.mMesh.AddCellData(kVoxelOccupancyName, std::move(blocks));
        }
        return result;
    }

    // A selective fill keeps a subset of the cells. Building the full lattice and
    // then subsetting it would cost the memory the budget check exists to avoid,
    // so the kept cells are emitted directly and the points they reference are
    // compacted in ascending order -- surface.cpp's used/remap pattern, which is
    // what makes the output independent of which thread found which cell.
    std::int64_t nkept = 0;
    for (std::int64_t c = 0; c < ncells; ++c)
        nkept += occupied[static_cast<std::size_t>(c)] ? 1 : 0;
    result.mNumOccupied = nkept;
    result.mMesh = vox_build_subset(spec, occupied, nkept);
    if (rOptions.mAttachOccupancy && nkept > 0) {
        NDArray occ = NDArray::Uninit(DType::Int64, {static_cast<std::size_t>(nkept)});
        std::int64_t* dst = occ.As<std::int64_t>();
        parallel_for_bw(static_cast<std::size_t>(nkept), [&](std::size_t c) { dst[c] = 1; });
        std::vector<NDArray> blocks;
        blocks.push_back(std::move(occ));
        result.mMesh.AddCellData(kVoxelOccupancyName, std::move(blocks));
    }
    return result;
}

}  // namespace meshioplusplus
